helper/ff_functions.py

```diff
@@ -3,21 +3,111 @@
+    hardcoded as "ntuple". The function assumes that the wrapped function returns a ROOT RDataFrame.
+    The cache file is a ROOT file with a name derived from a hash of the function's arguments.
+
+    The decorator performs the following steps:
+        1. Checks if the cache file for the given arguments exists in the cache directory.
+        2. If the cache file exists, loads the RDataFrame from the file.
+        3. If the cache file does not exist, calls the wrapped function to process the RDataFrame,
+           saves the result to the cache file, and then reloads the RDataFrame from the file.
+
+    Args:
+        cache_dir (str): The directory where cached ROOT files will be stored. Defaults to "./.RDF_CACHE".
+
+    Returns:
+        Callable: A decorator that wraps a function which processes an RDataFrame.
+    """
+    def decorator(function: Callable) -> Callable:
+        @functools.wraps(function)
+        def wrapper(*args: Any, **kwargs: Any) -> ROOT.RDataFrame:
+            log = logging.getLogger(kwargs.get("logger") or function.__module__ + '.' + function.__name__)
+            tree_name = "ntuple"
+
+            if 'rdf' in kwargs:
+                base_rdf = kwargs['rdf']
+                key_args = {k: v for k, v in kwargs.items() if k != 'rdf'}
+            elif args:
+                base_rdf = args[0]
+                key_args = {'__args__': args[1:], **kwargs}
+            else:
+                raise ValueError("Could not find RDataFrame argument ('rdf').")
+
+            key_args['_var_'] = func.CachingKeyHelper.get_assignment_variable()
+            caller_frame = inspect.stack()[1]
+            caller_info = f"{caller_frame.function}@{caller_frame.filename}:{caller_frame.lineno}"
+            key_args["_caller_"] = caller_info
+
+            if "logger" in key_args:
+                del key_args["logger"]
+
+            os.makedirs(cache_dir, exist_ok=True)
+
+            cache_hash = func.CachingKeyHelper.generate_key((), func.CachingKeyHelper.make_hashable(key_args))
+            cache_filepath = os.path.join(cache_dir, f"{cache_hash}.root")
+
+            if os.path.exists(cache_filepath) and func.RuntimeVariables.USE_CACHED_INTERMEDIATE_STEPS:
+                log.info(f"Using existent filtered Rdf: {cache_filepath}")
+                return ROOT.RDataFrame(tree_name, cache_filepath)
+
+            log.info(f"Creating filtered Rdf under: {cache_filepath}")
+
+            cols = [str(c) for c in base_rdf.GetColumnNames()]
+            filtered_rdf = function(*args, **kwargs)
+
+            if filtered_rdf.Count().GetValue() == 0:
+                log.warning("Filter resulted in zero events. Creating an empty snapshot with the correct schema.")
+                f = ROOT.TFile(cache_filepath, "RECREATE")
+                tree = ROOT.TTree(tree_name, tree_name)
+                for c in cols:
+                    arr = ROOT.std.vector('float')()
```

Comment:
Does it have to be `ROOT.std.vector`? What I want to say is, I would find `ROOT.RVec` more consistent or not use a root for the vector at all if possible.